### DISCUSSION TITLE
What does this exception during refresh mean: No column-based properties specified for refresh operation.

### DISCUSSION THREAD

**Discussion by pankajkumar229** [sqlalchemy/sqlalchemy#9287] [Usage Questions]
I am trying to refresh some ORM attributes but getting this error. Not sure how to figure out column based properties. Should relation based attributes not be used in refresh?

**Comment by zzzeek** [comment]
that issue has been fixed in 2.0, https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-b58310d7558f78d57171e8df4c4c3aae .

if using 1.4, add the primary key columns to the list of attributes, see workaround at https://github.com/sqlalchemy/sqlalchemy/issues/8703#issuecomment-1288201336

**Comment by pankajkumar229** [comment]
![Screen Shot 2023-02-12 at 1 35 44 PM](https://user-images.githubusercontent.com/1482916/218338604-e8bc54c7-9f3d-4ba1-a612-af73659ce7b8.png)

Hello, 

Thanks for your prompt replies. I tried using the method but cannot get the attributes to load. It seems refresh is not loading the attributes. Please see screenshot. I was expecting "courses" attribute to get loaded to the "next_obj" object which is of type "User" but it still shows error after refresh.

**Reply by zzzeek** [reply]
your screenshot doesn't really show me anything.     can you clarify what "the method" is ?  did you turn on `echo='debug'` on your engine so you can see what SQL is emitted when you call upon refresh?

**Comment by pankajkumar229** [comment]
Oh I was using the workaround you suggested in the other thread:

```
def refresh(session, obj, attrs):
    insp = inspect(obj)
    attrs = set(attrs).union(
        [k.key for k in insp.mapper.column_attrs if k.expression.primary_key]
    )
    session.refresh(obj, attrs)
```

I used the same method to refresh objects. In the screenshot, we are trying to refresh {"id" , "courses"} on next_obj using "set(x).union(primary_keys)". But after the await returns, you can see in the debugger screenshot that next_obj's course attribute is still not populated (Hoping it will at least be None/Empty).

**Reply by zzzeek** [reply]
refresh only refreshes relationships that were eagerly loaded originally.  there are many reasons why your operation may be not working as expected which I can only comment on using isolated demonstration cases that I can run locally, no third party imports or external data needed.  See the [great mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22+is%3Aclosed) label for examples of test cases I can run locally.

**Reply by pankajkumar229** [reply]
Ahh that makes sense. I guess this method is not easy either. Do you think an execute after figuring out the relationship is possible?

**Reply by zzzeek** [reply]
yes, run a query for the parent object in question, indicate the eager loaders you wish on the object, and use [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing) on the query.  anything you indicate in your query (either 2.0 style select() or legacy Query) will be overwritten in the target object.

this is mentioned in the tip at https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.refresh

**Reply by CaselIT** [reply]
> refresh only refreshes relationships that were eagerly loaded originally. there are many reasons why your operation may be not working as expected which I can only comment on using isolated demonstration cases that I can run locally, no third party imports or external data needed. See the [great mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22+is%3Aclosed) label for examples of test cases I can run locally.

Maybe it should warn (and error in a future release) when passing relationship keys that are not refreshable because they were not eagerly loaded?

**Reply by zzzeek** [reply]
or we can make it load.  I've opened #9298 which illustrates both approaches.   needs an additional parameter passed through from session.refresh() and some tests.

**Reply by CaselIT** [reply]
works for me too. What I think make sense avoid is silently failing behavior, since it's very hard and frustrating to catch
